### PR TITLE
Improve main donation form

### PR DESCRIPTION
**Pull request by marekrozmus**
Added some improvements.

### CODE REVIEW

File: components/MainDonationForm.js [hunk 1]
```diff
@@ -19,11 +20,35 @@ const AMOUNTS = [
...
+  React.useEffect(() => {
+    if (isAnyoneInNeed) {
+      setCause('Anyone in Need');
+    } else {
+      const causes = CAUSES.filter(({ id }) => selectedCauses[id]).map(
+        ({ label }) => label
+      );
+      setCause(causes.join(', '));
+    }
+  }, [isAnyoneInNeed, selectedCauses]);
```
Comment (audreyfeldroy) [inline components/MainDonationForm.js:51]: I like this! Wow, you've shortened the code substantially.

File: components/MainDonationForm.js [hunk 2]
```diff
@@ -32,162 +57,62 @@ export default function MainDonationForm() {
...
-    // If the "Anyone" toggle is on, or if all causes are selected,
-    // summarize all causes as "Anyone in Need"
-    if (isAnyoneInNeedToggled || causes.length === selectedCauses.length) {
-      selectedCauses = [' Anyone in Need'];
-    }
-
-    if (selectedCauses.length === 0) {
-      showModal('Please select a cause for donation');
+    if (!amount) {
+      showModal('Please choose an amount to give');
```
Comment (audreyfeldroy) [inline components/MainDonationForm.js:72]: I like the more generic falsy check for `amount`.

File: components/form/CheckboxFormControl.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      onChange={handleChange}
+      type="checkbox"
+    />
+  );
+}
+
+function CheckboxFormControl({ id, label, isChecked, isToggle, onChange }) {
+  return (
+    <div className="form-control">
+      {label ? (
```
Comment (audreyfeldroy) [inline components/form/CheckboxFormControl.js:22]: Interesting, what's the scenario where `CheckboxFormControl` has no label?

File: components/MainDonationForm.js [hunk 4]
```diff
@@ -32,162 +57,62 @@ export default function MainDonationForm() {
...
-                onChange={handleClick}
-                type="checkbox"
-                value="Anyone in Need"
+          <CheckboxFormControl
+            id="cause"
+            isChecked={isAnyoneInNeed}
+            isToggle
+            label="Anyone in Need"
+            onChange={setIsAnyoneInNeed}
+          />
```
Comment (audreyfeldroy) [inline components/MainDonationForm.js:95]: Beautiful, this is so nice and clean.

File: components/MainDonationForm.js [hunk 5]
```diff
@@ -32,162 +57,62 @@ export default function MainDonationForm() {
...
+            ))}
           <SelectFormControl
             id="amount"
             onChange={setAmount}
             options={AMOUNTS}
             value={amount}
           />
-          <button className="btn btn-primary" type="submit">
+          <button
+            className={`btn btn-primary ${isPending ? 'loading' : ''}`}
```
Comment (audreyfeldroy) [inline components/MainDonationForm.js:113]: Nice touch with adding the `loading` state animation.

File: components/form/SelectFormControl.js [hunk 6]
```diff
@@ -33,11 +33,17 @@ SelectFormControl.propTypes = {
...
+        PropTypes.string.isRequired,
+        PropTypes.number.isRequired,
+      ]),
     })
   ),
-  value: PropTypes.string.isRequired,
+  value: PropTypes.oneOfType([
+    PropTypes.string.isRequired,
+    PropTypes.number.isRequired,
+  ]),
```
Comment (audreyfeldroy) [inline components/form/SelectFormControl.js:46]: Nice.

**Review by audreyfeldroy — APPROVED** [review]
This is awesome @marekrozmus.

File: components/MainDonationForm.js [hunk 7]
```diff
@@ -19,11 +20,35 @@ const AMOUNTS = [
   { value: 25000, label: '$25,000' },
 ];
 
+const CAUSES = [
+  { id: 1, label: 'Students in Need' },
```
Comment (pydanny) [inline components/MainDonationForm.js:24]:
```suggestion
  { value: 1, label: 'Students in Need' },
```
The API for `SelectFormControl` looks for `value`, not `id`.

File: components/MainDonationForm.js [hunk 8]
```diff
@@ -19,11 +20,35 @@ const AMOUNTS = [
...
+  { id: 2, label: 'People of Color in Need' },
```
Comment (pydanny) [inline components/MainDonationForm.js:25]:
```suggestion
  { value: 2, label: 'People of Color in Need' },
```
The API for `SelectFormControl` looks for `value`, not `id`.

File: components/MainDonationForm.js [hunk 9]
```diff
@@ -19,11 +20,35 @@ const AMOUNTS = [
...
+  { id: 3, label: 'Immigrants in Need' },
```
Comment (pydanny) [inline components/MainDonationForm.js:26]:
```suggestion
  { value: 3, label: 'Immigrants in Need' },
```
The API for `SelectFormControl` looks for `value`, not `id`.

File: components/MainDonationForm.js [hunk 10]
```diff
@@ -19,11 +20,35 @@ const AMOUNTS = [
...
+  { id: 4, label: 'Seniors in Need' },
```
Comment (pydanny) [inline components/MainDonationForm.js:27]:
```suggestion
  { value: 4, label: 'Seniors in Need' },
```
The API for `SelectFormControl` looks for `value`, not `id`.

**Review by pydanny — CHANGES_REQUESTED** [review]
Looks solid to me, just including suggestions to address changes to `SelectFormControl` introduced in 0834244ea73c1c41eacab732170bb83e05b7cb64.

File: components/MainDonationForm.js (hunk 10 above)
Comment (marekrozmus) [reply components/MainDonationForm.js:27]: The `CAUSES` is used by `CheckboxFormControl`, not by `SelectFormControl`. For checkboxes I just use `id` to keep information about which one is selected. Or am I missing something 🙂 ?

**Comment by marekrozmus** [comment]
Please do not merge yet - I want to add functionality that when all checkboxes are selected then the "all in need" selector is switched on.

File: components/form/CheckboxFormControl.js (hunk 3 above)
Comment (marekrozmus) [reply components/form/CheckboxFormControl.js:22]: you are right - now I cannot think of such situation 🙂 - I will remove that

**Comment by marekrozmus** [comment]
@audreyfeldroy this one is read to merge

**Comment by audreyfeldroy** [comment]
Merged, thank you @marekrozmus and @pydanny!

**Pull request merged**